### PR TITLE
Add support for instrumentation scope attributes (Meter tags)

### LINKED ISSUE #4563: Export Instrumentation Scope attributes for a Meter (requires DiagnosticSource v8)
System.Diagnostics.DiagnosticSource v8.0 is introducing the ability to [instantiate a `Meter` with tags](https://github.com/dotnet/runtime/blob/a2656afabd9057dd0ccabcecd411589a4afaaafa/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs#L62)

This satisfies the requirement from the spec to provide instrumentation scope attributes. See:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#get-a-meter.

Once we bump our dependency to DiagnosticSource v8.0 we will need to update exporters to export these tags.

### PR DESCRIPTION
**Pull request by cijothomas**
Fixes #4563

Did not touch Prometheus exporters in this PR, better to wait till https://github.com/open-telemetry/opentelemetry-dotnet/pull/5086 is merged.

## Changes

Please provide a brief description of the changes here.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry/Metrics/Metric.cs [hunk 1]
```diff
@@ -207,6 +207,11 @@ internal static readonly HashSet<(string, string)> DefaultHistogramBoundLongMapp
     /// </summary>
     public string MeterVersion => this.InstrumentIdentity.MeterVersion;
 
+    /// <summary>
+    /// Gets the attributes (tags) for the metric stream.
+    /// </summary>
+    public IEnumerable<KeyValuePair<string, object?>> MeterTags => this.InstrumentIdentity.MeterTags;
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/Metric.cs:213]: Are meter tags expected to be unaffected by Views?

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Metric.cs:213]: yes. just like Meter name and version.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/MetricItemExtensions.cs [hunk 2]
```diff
@@ -1,418 +1,434 @@
...
+        {
+            metrics = new OtlpMetrics.ScopeMetrics
+            {
+                Scope = new OtlpCommon.InstrumentationScope
+                {
+                    Name = name, // Name is enforced to not be null, but it can be empty.
+                    Version = version ?? string.Empty, // NRE throw by proto
+                },
+            };
+            metrics.Scope.Attributes.Clear();
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/MetricItemExtensions.cs:101]: nit: not needed.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/MetricItemExtensions.cs [hunk 3]
```diff
@@ -1,418 +1,434 @@
...
+            AddAttributes(meterTags, metrics.Scope.Attributes);
+        }
+        else
+        {
+            metrics.Scope.Name = name;
+            metrics.Scope.Version = version ?? string.Empty;
+            metrics.Scope.Attributes.Clear();
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/MetricItemExtensions.cs:108]: This could be moved to line `82` to keep all clean up together?

File: src/OpenTelemetry/Metrics/Metric.cs (hunk 1 above)
Comment (utpilla) [reply src/OpenTelemetry/Metrics/Metric.cs:213]: Is this coming from the spec? We are exporting Meter tags are regular tags as well, right?

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Metric.cs:213]: This is part of spec.
Meter tags are separate from regular tags. Metertags are not part of aggregation. It is at the same level as Meter.Name,Version, forming the "instrumentation scope" from the spec.

File: src/OpenTelemetry/Metrics/MetricStreamIdentity.cs [hunk 4]
```diff
@@ -101,6 +104,8 @@ public MetricStreamIdentity(Instrument instrument, MetricStreamConfiguration? me
 
     public string MeterVersion { get; }
 
+    public IEnumerable<KeyValuePair<string, object?>> MeterTags { get; }
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/MetricStreamIdentity.cs:107]: Should we embrace nullability here?

```diff
    public MetricStreamIdentity(Instrument instrument, MetricStreamConfiguration? metricStreamConfiguration)
    {
-        this.MeterTags = instrument.Meter.Tags ?? Enumerable.Empty<KeyValuePair<string, object?>>();
+        this.MeterTags = instrument.Meter.Tags;
    }

-    public IEnumerable<KeyValuePair<string, object?>> MeterTags { get; }
+    public IEnumerable<KeyValuePair<string, object?>>? MeterTags { get; }
```

Thinking the way it is currently there isn't a super cheap way to skip processing for the common case where it isn't set.

```csharp
    AddAttributes(meterTags, metrics.Scope.Attributes); // pay for a call

    private static void AddAttributes(IEnumerable<KeyValuePair<string, object>> meterTags, RepeatedField<OtlpCommon.KeyValue> attributes)
    {
        foreach (var tag in meterTags) // Pay to call GetEnumerator/MoveNext/check result
        {
            if (OtlpKeyValueTransformer.Instance.TryTransformTag(tag, out var result))
            {
                attributes.Add(result);
            }
        }
    }
```

VS:

```csharp
    if (meterTags != null) // Pay for a null check only
    {
       AddAttributes(meterTags, metrics.Scope.Attributes);
    }
```

File: src/OpenTelemetry/Metrics/Metric.cs (hunk 1 above)
Comment (utpilla) [reply src/OpenTelemetry/Metrics/Metric.cs:213]: Or is it up to the exporter/backend to decide how it wants to treat them? In any case, there should be an official stance documented for how [this](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/metrics/customizing-the-sdk#select-specific-tags) affects or does not affect Meter tags.

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Metric.cs]: mm.. The name "tags/attributes" is what is confusing? There are tags/attributes on Resource level, then again on Meter level, and then on each recording - all are independent. The ones on recording are used for Aggregationg, the ones on meter are used to populate InstrumentaionScope, and one on Resource are used to.. well populate Resource...

Do we need to do anything about this? I can think of mentioning this in our best-practices doc so users can pick the right choice based on the needs and avoid mistakes/less-ideal solutions like [this](https://github.com/open-telemetry/opentelemetry-dotnet/discussions/5088).

(I intentionally did not mention anything about the Instrument level tags .NET 8 added... That is not part of spec, and we'll need to discuss how to deal with that later. It could be merged with with Meter tags (i.e won't affect aggreagation) OR merged with Tags provided at measurement time (affecting aggregation)...That would make things more interesting , lets worry about it later.

Comment (alanwest) [reply src/OpenTelemetry/Metrics/Metric.cs]: I discussed this awhile back with @jack-berg from Java. The conclusion we came to is that it would be incorrect for views to have any affect over scope attributes.

Somewhat similarly, views do not affect resource attributes, and for good reason because if a view could affect resource attributes this would open the door for an SDK instance to emit multiple distinct resources which is not allowed by the spec.

More generally and with regards to scope attributes, the attribute keys defined by a view only govern the attributes supplied by individual measurements - i.e., the attributes influencing aggregation as @cijothomas notes.

Comment (alanwest) [reply src/OpenTelemetry/Metrics/Metric.cs]: > I intentionally did not mention anything about the Instrument level tags .NET 8 added...

This is essentially "bound" instruments and to my knowledge has gotten some renewed interest for reintroducing in the spec.

Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/Metric.cs]: It isn't a blocker for me but I would personally feel better if the spec was clarified to say something like: "Views are NOT able to select/drop attributes inherited via Resource or Instrumentation Scope."

Resource users can control. But instrumentation scope you are kind of at the mercy of the author to not add a bunch of noise. If they do, you will have to pay for it 💰 For that reason I can imagine the spec one day becoming opinionated about how to drop stuff added through the instrumentation scope (may or may not be through the View API or something else).

File: test/OpenTelemetry.Tests/Metrics/MetricApiTestsBase.cs [hunk 5]
```diff
@@ -178,6 +178,36 @@ public void MetricDescriptionIsExportedCorrectly(string description)
         Assert.Equal(description ?? string.Empty, metric.Description);
     }
 
+    [Fact]
```
Comment (alanwest) [inline test/OpenTelemetry.Tests/Metrics/MetricApiTestsBase.cs:181]: Just to clarify and document the behavior, consider adding a test with duplicate meters.

```csharp
using var meter1 = new Meter("MyMeter", "1.0", new List<KeyValuePair<string, object>> { new("Key1", "Value1") });
using var meter2 = new Meter("MyMeter", "1.0", new List<KeyValuePair<string, object>> { new("Key2", "Value2") });
```

I think what will happen here is that measurements from an identical instrument from each of the meters will be exported with the tags from meter1, correct?

File: src/OpenTelemetry/Metrics/Metric.cs (hunk 1 above)
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Metric.cs]: > If they do, you will have to pay for it 

That really depends on the backend! Good chance most backends are just dropping it today (given its a new feature). Or could be opt-in basis as well! The only cost is the cost of serializing+network - but instrumentation scope is not repeated for each metricpoint, so not that bad.

File: src/OpenTelemetry.Exporter.Console/ConsoleMetricExporter.cs [hunk 6]
```diff
@@ -75,6 +75,15 @@ public override ExportResult Export(in Batch<Metric> batch)
 
             this.WriteLine(msg.ToString());
 
+            foreach (var meterTag in metric.MeterTags)
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.Console/ConsoleMetricExporter.cs:78]: Update CHANGELOG for ConsoleExporter as well.

File: test/OpenTelemetry.Tests/Metrics/MetricApiTestsBase.cs (hunk 5 above)
Comment (cijothomas) [reply test/OpenTelemetry.Tests/Metrics/MetricApiTestsBase.cs:181]: added tests for this explicitly.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/CHANGELOG.md [hunk 7]
```diff
@@ -15,6 +15,9 @@
   accepts a `name` parameter to support named options.
   ([#4916](https://github.com/open-telemetry/opentelemetry-dotnet/pull/4916))
 
+* Add support for Instrumentation Scope Attributes (i.e Meter Tags), fixing
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/CHANGELOG.md:18]: Add a link to the `Meter.Tags` doc: https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.metrics.meter.tags?view=net-8.0

File: src/OpenTelemetry/CHANGELOG.md [hunk 8]
```diff
@@ -58,6 +58,9 @@
   implementationFactory)`.
   ([#4916](https://github.com/open-telemetry/opentelemetry-dotnet/pull/4916))
 
+* Add support for Instrumentation Scope Attributes (i.e Meter Tags), fixing
+  issue [#4563](https://github.com/open-telemetry/opentelemetry-dotnet/issues/4563).
```
Comment (utpilla) [inline src/OpenTelemetry/CHANGELOG.md:62]: nit: Add PR link for easier access.

File: src/OpenTelemetry/Metrics/Metric.cs (hunk 1 above)
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Metric.cs]: > > I intentionally did not mention anything about the Instrument level tags .NET 8 added...
> 
> This is essentially "bound" instruments and to my knowledge has gotten some renewed interest for reintroducing in the spec.

The topic of bound instrument is precisely why I started looking into this! 😄 (OTel Rust has solid need for that, I was trying to prototype how it'd look in a more familiar code-base)

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/MetricItemExtensions.cs [hunk 9]
```diff
@@ -97,11 +97,21 @@ internal static OtlpMetrics.ScopeMetrics GetMetricListFromPool(string name, stri
...
+                AddAttributes(meterTags, metrics.Scope.Attributes);
+            }
         }
         else
         {
             metrics.Scope.Name = name;
             metrics.Scope.Version = version ?? string.Empty;
+            if (meterTags != null)
+            {
+                metrics.Scope.Attributes.Clear();
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/MetricItemExtensions.cs:112]: I think we should clear these attributes in the `Return` method to stay consistent with how we clear `ScopeMetrics.Metrics`.

https://github.com/open-telemetry/opentelemetry-dotnet/blob/324ab61fa190880222468992f29844df2559ac24/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/MetricItemExtensions.cs#L72-L85

File: src/OpenTelemetry/Metrics/Metric.cs (hunk 1 above)
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Metric.cs]: All okay to resolve this? 
I will sent a follow up for update metric best practices to talk about Meter.Tags, and also highlight the importance of leveraging Resources more. 
Not planning to do any doc change for the View part.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/MetricItemExtensions.cs (hunk 9 above)
Comment (cijothomas) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/MetricItemExtensions.cs:112]: Thanks. https://github.com/open-telemetry/opentelemetry-dotnet/pull/5089#discussion_r1408254794 is suggesting the same thing I believe.

**Review by alanwest — APPROVED** [review]

File: src/OpenTelemetry/Metrics/Metric.cs [hunk 10]
```diff
@@ -207,6 +207,11 @@ internal static readonly HashSet<(string, string)> DefaultHistogramBoundLongMapp
     /// </summary>
     public string MeterVersion => this.InstrumentIdentity.MeterVersion;
 
+    /// <summary>
+    /// Gets the attributes (tags) for the metric stream.
+    /// </summary>
+    public IEnumerable<KeyValuePair<string, object?>>? MeterTags => this.InstrumentIdentity.MeterTags;
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/Metric.cs:213]: Only expose this public API only on .NET 8 and above? It's not applicable to the older targets.

Comment (utpilla) [reply src/OpenTelemetry/Metrics/Metric.cs:213]: Ahh never mind! This is coming from DS 8.0.

File: test/OpenTelemetry.Tests/Metrics/MetricApiTestsBase.cs [hunk 11]
```diff
@@ -178,6 +178,83 @@ public void MetricDescriptionIsExportedCorrectly(string description)
...
+
+        var counter = meter.CreateCounter<long>("name1");
+        counter.Add(10);
+        meterProvider.ForceFlush(MaxTimeToAllowForFlush);
+        Assert.Single(exportedItems);
+        var metric = exportedItems[0];
+        Assert.Equal(meterName, metric.MeterName);
+        Assert.Equal(meterVersion, metric.MeterVersion);
+
+        bool containsMeterTags = metric.MeterTags.Any(kvp =>
```
Comment (utpilla) [inline test/OpenTelemetry.Tests/Metrics/MetricApiTestsBase.cs:206]: Use `Assert.Single` with `Where` instead.

https://github.com/open-telemetry/opentelemetry-dotnet/blob/324ab61fa190880222468992f29844df2559ac24/test/OpenTelemetry.Tests/Resources/ResourceBuilderTests.cs#L29

File: test/OpenTelemetry.Tests/Metrics/MetricApiTestsBase.cs [hunk 12]
```diff
@@ -178,6 +178,83 @@ public void MetricDescriptionIsExportedCorrectly(string description)
...
+            .AddInMemoryExporter(exportedItems));
+
+        var counter1 = meter1.CreateCounter<long>("my-counter");
+        counter1.Add(10);
+        var counter2 = meter2.CreateCounter<long>("my-counter");
+        counter2.Add(10);
+        meterProvider.ForceFlush(MaxTimeToAllowForFlush);
+
+        // The instruments differ only in the Meter.Tags, which is not an identifying property.
+        // The first instrument's Meter.Tags is exported.
```
Comment (utpilla) [inline test/OpenTelemetry.Tests/Metrics/MetricApiTestsBase.cs:245]: Add an explicit Assert statement to verify that the second Meter's tags are not exported.

File: test/OpenTelemetry.Tests/Metrics/MetricApiTestsBase.cs [hunk 13]
```diff
@@ -178,6 +178,83 @@ public void MetricDescriptionIsExportedCorrectly(string description)
...
+        // It is considered a user-error to create Meters with same name,version but with
+        // different tags. TODO: See if we can emit an internal log about this.
+        Assert.Single(exportedItems);
+        var metric = exportedItems[0];
```
Comment (utpilla) [inline test/OpenTelemetry.Tests/Metrics/MetricApiTestsBase.cs:249]: Add an Assert statement to verify that the measurement value exported equals 20. We want to verify that both the counters are tracked using the same `Metric`.

**Review by utpilla — APPROVED** [review]

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/MetricItemExtensions.cs (hunk 3 above)
Comment (cijothomas) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/MetricItemExtensions.cs]: https://github.com/open-telemetry/opentelemetry-dotnet/issues/5090 Will address this separately to unblock the merge and release asap.
